pkg/volume/emptydir/empty_dir.go

```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
 	return ed.mounter.Mount("nodev", dir, "hugetlbfs", []string{pageSizeMountOption})
 }
 
-// getPageSizeMountOptionFromPod retrieves pageSize mount option from Pod's resources
-// and validates pageSize options in all containers of given Pod.
-func getPageSizeMountOptionFromPod(pod *v1.Pod) (string, error) {
+// getPageSizeMountOption retrieves pageSize mount option from Pod's resources
+// and medium and validates pageSize options in all containers of given Pod.
+func getPageSizeMountOption(medium v1.StorageMedium, pod *v1.Pod) (string, error) {
 	pageSizeFound := false
 	pageSize := resource.Quantity{}
-	// In some rare cases init containers can also consume Huge pages.
-	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
-	for _, container := range containers {
+
+	var mediumPageSize resource.Quantity
+	var err error
+	if medium != v1.StorageMediumHugePages {
+		// medium is: Hugepages-<size>
+		mediumPageSize, err = v1helper.HugePageSizeFromMedium(medium)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// In some rare cases init containers can also consume Huge pages
+	for _, container := range append(pod.Spec.Containers, pod.Spec.InitContainers...) {
```

Comment:
In general this looks nice, but I am not sure if I like the fact that we do pod-spec validation here and not when validating the pod spec during creating of the pod.

If we "skip" the validation here we can simplify the code quite a bit.


Currently it is possible to successfully schedule a pod with a huge page volume mount without a huge page request. That will successfully schedule the pod, but it will stay in `ContainerCreating` becuase mounting the volume fails:

```
  Warning  FailedMount  48s (x8 over 111s)  kubelet, tar       MountVolume.SetUp failed for volume "hugepage1" : hugePages storage requested, but there is no resource request for huge pages
  Warning  FailedMount  48s (x8 over 111s)  kubelet, tar       MountVolume.SetUp failed for volume "hugepage" : hugePages storage requested, but there is no resource request for huge pages
  Warning  FailedMount  48s (x8 over 111s)  kubelet, tar       MountVolume.SetUp failed for volume "hugepage2" : hugePages storage requested, but there is no resource request for huge pages
```

The KEP states that pods like these are "invalid", not that "their volume mounts should fail", so i think failing during validation would be more correct.

Or am I missing something?

Reply:
This is more of a discussion that something that should be fixed in the PR, so i would love some thoughts from you. :smile:

Reply:
@odinuge Thank you for the review. I agree with your point that validation of the pod should happen earlier. I just followed the current code in a hope that it's implemented correctly. The validation was there, I've just extended it to cover new functionality.

I'll definitely look if it can be done earlier.

Reply:
Nice.

I think we can keep this now in order to make the changes here small, and move the validation in another PR.

Reply:
sgtm. will do.